presto-main/src/main/java/com/facebook/presto/sql/planner/MergeJoinWithGroupedExecutionOptimizer.java

```diff
@@ -0,0 +1,94 @@
+import com.facebook.presto.spi.eventlistener.PlanOptimizerInformation;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.MergeJoinNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.facebook.presto.sql.planner.plan.SortNode;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.planPrinter.PlanPrinter.jsonFragmentPlan;
+
+public class MergeJoinWithGroupedExecutionOptimizer
+{
+    public static SubPlan optimizeMergeJoinWithGroupedExecution(Session session, SubPlan subPlan, Metadata metadata)
+    {
+        if (!subPlan.getFragment().getStageExecutionDescriptor().isStageGroupedExecution()) {
+            return subPlan;
+        }
+        PlanNode original = subPlan.getFragment().getRoot();
+
+        PlanNode rewritten = SimplePlanRewriter.rewriteWith(new Rewriter(), original);
+        if (rewritten == original) {
+            return subPlan;
+        }
+
+        session.getOptimizerInformationCollector().addInformation(new PlanOptimizerInformation(MergeJoinWithGroupedExecutionOptimizer.class.getSimpleName(), true, Optional.empty()));
+        PlanFragment originalFragment = subPlan.getFragment();
+
+        return new SubPlan(
+                new PlanFragment(
+                        originalFragment.getId(),
+                        rewritten,
+                        originalFragment.getVariables(),
+                        originalFragment.getPartitioning(),
+                        originalFragment.getTableScanSchedulingOrder(),
+                        originalFragment.getPartitioningScheme(),
+                        originalFragment.getStageExecutionDescriptor(),
+                        originalFragment.isOutputTableWriterFragment(),
+                        originalFragment.getStatsAndCosts(),
+                        Optional.of(jsonFragmentPlan(rewritten, originalFragment.getVariables(), originalFragment.getStatsAndCosts(), metadata.getFunctionAndTypeManager(), session))),
+                subPlan.getChildren());
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        @Override
+        public PlanNode visitMergeJoin(MergeJoinNode node, RewriteContext<Void> rewriteContext)
+        {
+            if (canRemoveMergingExchange(node.getLeft()) && canRemoveMergingExchange(node.getRight())) {
+                return node.replaceChildren(ImmutableList.of(node.getLeft().getSources().get(0), node.getRight().getSources().get(0)));
+            }
+            return node;
+        }
+
+        private boolean canRemoveMergingExchange(PlanNode node)
+        {
+            return isMergingExchange(node) && node.getSources().size() == 1 && !containsSortOrExchange(node.getSources().get(0));
```

Comment:
are these conditions sufficient?  Do we need to check also that there are no remote source nodes (you can have a remote source node with grouped execution with broadcast join). Are there any other conditions we need to check?